File: CPP/ByTier/D/2093_D_Skibidi_Table.cpp
// Problem: CF 2093 D - Skibidi Table
// https://codeforces.com/contest/2093/problem/D

/*
D. Skibidi Table
Algorithm: 
    This problem uses a recursive approach to map coordinates to numbers and vice versa 
    in an N x N matrix filled in a specific order (Z-order or Morton order).
    
    The filling process:
        1. Divide the square matrix into four equal quadrants.
        2. Fill them in this order: top-left, bottom-right, bottom-left, top-right.
        3. Recursively apply the same logic to each quadrant until 2x2 matrices are reached,
           which are filled as: top-left=1, bottom-right=2, bottom-left=3, top-right=4.

    Time Complexity: O(n * q) per test case where n is the size of the matrix (2^n x 2^n), 
                     and q is the number of queries.
                     Each query involves traversing a path from the full matrix down to 
                     a cell or back from the cell to the matrix level, taking at most O(n).
    Space Complexity: O(1) - only using basic integer variables and no additional data structures.

*/

#include <iostream>
#include <string>

using namespace std;

void solve() {
  int n, q;
  cin >> n >> q;
  while (q--) {
    string s;
    cin >> s;
    if (s[0] == '-') { // Query type: find number at coordinates x,y
      int x, y;
      cin >> x >> y;
      long long ans = 0;
      int m = n;
      int ad = 1 << (n - 1); // Half the size of current square being processed
      while (m--) {
        ans <<= 2; // Shift left by 2 bits to accommodate next quadrant code (0-3)
        if (x > ad && y > ad) // Bottom right quadrant
          x -= ad, y -= ad, ans += 1;
        if (x > ad && y <= ad) // Bottom left quadrant
          x -= ad, ans += 2;
        if (x <= ad && y > ad) // Top right quadrant
          y -= ad, ans += 3;
        ad >>= 1; // Move to next smaller square
      }
      ans++; // Adjust for 1-based numbering of final element
      cout << ans << '\n';
    } else { // Query type: find coordinates for a given number d
      long long d;
      cin >> d;
      d--; // Convert to 0-indexed
      int x = 1, y = 1;
      int ad = 1; // Current square size (initially 1x1)
      while (d) {
        int now = d & 3; // Extract last two bits to get quadrant code
        if (now == 1) // Bottom right quadrant
          y += ad, x += ad;
        if (now == 2) // Bottom left quadrant
          x += ad;
        if (now == 3) // Top right quadrant
          y += ad;
        d >>= 2; // Move to parent level
        ad <<= 1; // Double square size
      }
      cout << x << ' ' << y << '\n';
    }
  }
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0);
  cout.tie(0);
  int t;
  cin >> t;
  while (t--) {
    solve();
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/D/2093_D_Skibidi_Table.py
# Problem: CF 2093 D - Skibidi Table
# https://codeforces.com/contest/2093/problem/D

"""
D. Skibidi Table

Algorithms/Techniques:
- Recursive decomposition of a 2^n x 2^n table into four quadrants.
- Each quadrant is recursively filled in a specific order (top-left, bottom-right, bottom-left, top-right).
- The mapping between coordinates and numbers, or vice versa, uses bit manipulation.

Time Complexity: O(q * n), where q is the number of queries and n is the size parameter.
Space Complexity: O(1) excluding output storage.

This problem involves navigating a recursively filled 2^n x 2^n table. 
The approach leverages bit manipulation to determine how a cell's coordinates or a number maps 
to its position in the recursive structure.
"""

def __Main__():
    T = int(input())
    outs = []
    for _ in range(T):
        n = int(input())
        for _ in range(int(input())):
            f = list(input().split())
            if f[0] == "->":
                x, y = int(f[1]) - 1, int(f[2]) - 1  # Convert to 0-based indexing
                d = 0
                for i in range(n):
                    nx, ny = x >> i & 1, y >> i & 1       # Extract i-th bit of x and y
                    nd = (nx * 2) ^ (ny * 3)             # Combine the bits according to specific rule
                    d |= nd << (2 * i)                   # Place result in appropriate position
                outs.append(str(d + 1))                  # Convert back to 1-based
            else:
                d = int(f[1]) - 1                        # Convert to 0-based
                x, y = 0, 0
                for i in range(n):
                    nd = d >> (2 * i) & 3                # Extract 2-bit chunk from d
                    nx, ny = (nd & 1) ^ (nd >> 1), nd & 1 # Decode bits back to x,y components
                    x |= nx << i                         # Build x coordinate
                    y |= ny << i                         # Build y coordinate
                outs.append(f"{x + 1} {y + 1}")          # Convert to 1-based indexing
    print("\n".join(map(str, outs)))


if __name__ == "__main__":
    __Main__()


# https://github.com/VaHiX/codeForces/